source/common/thread_local/thread_local_impl.h

```diff
@@ -57,6 +59,9 @@ class InstanceImpl : Logger::Loggable<Logger::Id::main>, public Instance {
 
   static thread_local ThreadLocalData thread_local_data_;
   std::vector<SlotImpl*> slots_;
+  // A set of index of freed slots.
+  absl::flat_hash_set<uint32_t> free_slot_indexes_;
```

Comment:
I think you can just use std::list here?

Reply:
if we don't care about the assertion around when we return the slot, we can use a list here. Please let me know if you prefer to change to list.